File: backend/test/domain/entitlementResolver.test.ts
import { describe, expect, it } from "vitest";
import { resolveEntitlement } from "../../src/domain/entitlementResolver.js";
import type { AccountEntitlementSnapshot, SkillDefinition } from "../../src/domain/types.js";

const now = new Date("2026-08-26T00:00:00Z");

function skill(overrides: Partial<SkillDefinition> = {}): SkillDefinition {
  return {
    id: "web.search",
    name: "Web Search",
    description: "test",
    category: "WEB",
    capabilities: [],
    requiredPermissions: [],
    requiredEntitlement: "FREE",
    usageCost: { value: 0, unit: "credits" },
    riskLevel: "LOW",
    requiresConfirmation: false,
    executesOnDevice: false,
    inputSchema: { requiredFields: [] },
    handler: async () => ({ kind: "success", output: {}, summary: "ok" }),
    ...overrides,
  };
}

describe("resolveEntitlement", () => {
  it("denies a FREE plan a PRO-required skill", () => {
    const snapshot: AccountEntitlementSnapshot = { accountId: "a1", plan: "FREE", trialExpiresAt: null, creditBalance: 100 };
    const decision = resolveEntitlement(snapshot, skill({ requiredEntitlement: "PRO" }), now);
    expect(decision).toEqual({ allowed: false, reason: "PLAN_TOO_LOW", upgradeTo: "PRO" });
  });

  it("allows an active trial to use a TRIAL-tier skill", () => {
    const snapshot: AccountEntitlementSnapshot = {
      accountId: "a1",
      plan: "TRIAL",
      trialExpiresAt: new Date(now.getTime() + 86_400_000),
      creditBalance: 100,
    };
    const decision = resolveEntitlement(snapshot, skill({ requiredEntitlement: "TRIAL" }), now);
    expect(decision).toEqual({ allowed: true });
  });

  it("falls back an expired trial to FREE and denies a TRIAL-tier skill", () => {
    const snapshot: AccountEntitlementSnapshot = {
      accountId: "a1",
      plan: "TRIAL",
      trialExpiresAt: new Date(now.getTime() - 86_400_000),
      creditBalance: 100,
    };
    const decision = resolveEntitlement(snapshot, skill({ requiredEntitlement: "TRIAL" }), now);
    expect(decision).toEqual({ allowed: false, reason: "TRIAL_EXPIRED", upgradeTo: "TRIAL" });
  });

  it("allows a PRO plan to use a PRO-required skill", () => {
    const snapshot: AccountEntitlementSnapshot = { accountId: "a1", plan: "PRO", trialExpiresAt: null, creditBalance: 100 };
    const decision = resolveEntitlement(snapshot, skill({ requiredEntitlement: "PRO" }), now);
    expect(decision).toEqual({ allowed: true });
  });

  it("denies insufficient credit balance even when the plan is sufficient", () => {
    const snapshot: AccountEntitlementSnapshot = { accountId: "a1", plan: "PRO", trialExpiresAt: null, creditBalance: 1 };
    const decision = resolveEntitlement(snapshot, skill({ usageCost: { value: 5, unit: "credits" } }), now);
    expect(decision).toEqual({ allowed: false, reason: "OUT_OF_CREDITS" });
  });
});

File: android/domain/src/test/kotlin/com/jarvismobile/domain/entitlement/EntitlementResolverTest.kt
package com.jarvismobile.domain.entitlement

import com.jarvismobile.domain.entity.AccountEntitlementSnapshot
import com.jarvismobile.domain.entity.EntitlementDecision
import com.jarvismobile.domain.entity.EntitlementDenialReason
import com.jarvismobile.domain.entity.EntitlementLevel
import com.jarvismobile.domain.entity.RiskLevel
import com.jarvismobile.domain.entity.SkillCategory
import com.jarvismobile.domain.entity.SkillDefinition
import com.jarvismobile.domain.entity.SkillHandler
import com.jarvismobile.domain.entity.SkillResult
import com.jarvismobile.domain.entity.UsageCost
import java.time.Instant
import java.time.temporal.ChronoUnit
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs

class EntitlementResolverTest {

    private val now = Instant.parse("2026-08-26T00:00:00Z")

    private fun skill(
        requiredEntitlement: EntitlementLevel = EntitlementLevel.FREE,
        usageCost: UsageCost = UsageCost.FREE,
    ) = SkillDefinition(
        id = "web.search",
        name = "Web Search",
        description = "test",
        category = SkillCategory.WEB,
        requiredEntitlement = requiredEntitlement,
        usageCost = usageCost,
        riskLevel = RiskLevel.LOW,
        requiresConfirmation = false,
        handler = SkillHandler { _, _ -> SkillResult.Success(emptyMap(), "ok") },
    )

    @Test
    fun `FREE plan is denied a PRO-required skill`() {
        val snapshot = AccountEntitlementSnapshot("acc-1", EntitlementLevel.FREE, trialExpiresAt = null, creditBalance = 100)
        val decision = EntitlementResolver.resolve(snapshot, skill(requiredEntitlement = EntitlementLevel.PRO), now)
        val denied = assertIs<EntitlementDecision.Denied>(decision)
        assertEquals(EntitlementDenialReason.PLAN_TOO_LOW, denied.reason)
        assertEquals(EntitlementLevel.PRO, denied.upgradeTo)
    }

    @Test
    fun `active trial allows a TRIAL-tier skill`() {
        val snapshot = AccountEntitlementSnapshot(
            "acc-1", EntitlementLevel.TRIAL, trialExpiresAt = now.plus(1, ChronoUnit.DAYS), creditBalance = 100,
        )
        val decision = EntitlementResolver.resolve(snapshot, skill(requiredEntitlement = EntitlementLevel.TRIAL), now)
        assertEquals(EntitlementDecision.Allowed, decision)
    }

    @Test
    fun `expired trial falls back to FREE and is denied a TRIAL-tier skill`() {
        val snapshot = AccountEntitlementSnapshot(
            "acc-1", EntitlementLevel.TRIAL, trialExpiresAt = now.minus(1, ChronoUnit.DAYS), creditBalance = 100,
        )
        val decision = EntitlementResolver.resolve(snapshot, skill(requiredEntitlement = EntitlementLevel.TRIAL), now)
        val denied = assertIs<EntitlementDecision.Denied>(decision)
        assertEquals(EntitlementDenialReason.TRIAL_EXPIRED, denied.reason)
    }

    @Test
    fun `PRO plan is allowed a PRO-required skill`() {
        val snapshot = AccountEntitlementSnapshot("acc-1", EntitlementLevel.PRO, trialExpiresAt = null, creditBalance = 100)
        val decision = EntitlementResolver.resolve(snapshot, skill(requiredEntitlement = EntitlementLevel.PRO), now)
        assertEquals(EntitlementDecision.Allowed, decision)
    }

    @Test
    fun `insufficient credit balance is denied even when the plan is sufficient`() {
        val snapshot = AccountEntitlementSnapshot("acc-1", EntitlementLevel.PRO, trialExpiresAt = null, creditBalance = 1)
        val decision = EntitlementResolver.resolve(snapshot, skill(usageCost = UsageCost(5)), now)
        val denied = assertIs<EntitlementDecision.Denied>(decision)
        assertEquals(EntitlementDenialReason.OUT_OF_CREDITS, denied.reason)
    }
}
